### PR TITLE
ResizableFrame: Make keyboard accessible

### LINKED ISSUE #51267: Site editor: frame resizing is not keyboard accessible
### Description

In the Site Editor, the new 'Frame resizing' introduced in https://github.com/WordPress/gutenberg/pull/49910 is not operable with the keyboard.

I do realize the new resizing functionality works pretty differently from the previous implementation. Anyways, the previous resizing feature used the `ResizeHandle` and, while not perfect, it was operable with the keyboard. The new functionality is not.

As a keyboard user, I can't resize the editor iframe.

I would like to kindly remind everyone that any feature must be operable with the keyboard. Or, there must be an alternative way to use a feature that is keyboard accessible. I couldn't find any way to resize the iframe by using the keyboard so I'd call this an accessibility regression.

Quickly scanning the related PR https://github.com/WordPress/gutenberg/pull/49910 I couldn't see any consideration about accessibility and keyboard interaction, which is a bit disheartening to be honest.

Additionally:
- The "Drag to resize" string [is not translatable](https://github.com/WordPress/gutenberg/blame/305d5e1820d209a66154f8dc48e73dc1e38db5e6/packages/edit-site/src/components/resizable-frame/index.js#L229).
- That string is used to render a HTML `title` attribute on the draggable div element. For a number of years, WordPress has been progressively removing title attributes from the UI because they serve litle purpose and are accessible to only a minority of users. See https://core.trac.wordpress.org/query?keywords=~title-attribute
We should never introduce new title attributes and use Tooltips instead.

### Step-by-step reproduction instructions

- Go to the Site Editor.
- Stay in the initial view, with the Navigation panel on the left and the editor canvas on the right.
- Hover between the panel and the canvas.
- Observe a 'drag handle' appears. While hovering, a title attribute shows up, with the text 'Drag to resize'.
- Use the keyboard and try to na igate to the drag handle.
- Observe the drag handle doesn't appear (it's not focusable anyways).
- Try to find an alternative way to resize the iframe by using only the keyboard.
- Observe there is no way to resize the iframe with the keyboard.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by mirka**
Fixes #51267 

## What?

Adds back keyboard operability to the Editor Canvas frame resizing handle.

In addition, the "Drag to resize" `title` attribute is now moved to a proper Tooltip and is translatable.

## Why?

The changes in #49910 introduced a major accessibility regression.

## How?

Make the resize handle focusable, and add keydown handlers for resizing.

## Testing Instructions

1. Open the [Site Editor](http://localhost:8888/wp-admin/site-editor.php).
2. Hit the Tab key until you focus on the resize handle.
3. Hit the Left/Right arrow keys to resize the frame. The Shift key works as a modifier.
4. Normal mouse drag resizing should also work as before.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/555336/4bc9177f-6492-40e3-83a0-809386496875

### CODE REVIEW

File: packages/edit-site/src/components/resizable-frame/style.scss
Comment (mirka) [inline packages/edit-site/src/components/resizable-frame/style.scss:69]: Unused styles 🧹

File: packages/edit-site/src/components/resizable-frame/style.scss
Comment (mirka) [inline packages/edit-site/src/components/resizable-frame/style.scss:39]: Resets user agent `button` styles.

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (mirka) [inline packages/edit-site/src/components/resizable-frame/index.js]: Needs accessibility feedback: Do we want some kind of focus manipulation after the frame is resized to full width? (cc @afercia @andrewhayward)

## To test

1. Focus on the resize handle.
2. Hit the left arrow key until the editor canvas is in full width (edit mode).
3. Hit the tab key to see where the focus is.

Comment out line 197 to see the default behavior.

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (mirka) [inline packages/edit-site/src/components/resizable-frame/index.js]: Makes `is-resizing` a proper CSS modifier for this element, instead of relying on the modifier on the parent element.

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (tellthemachines) [reply packages/edit-site/src/components/resizable-frame/index.js]: I think it would be good if focus behaved the same in this case as it does when moving to edit mode by pressing Enter after focusing the frame, where we land at the top of the editing canvas and Tab focuses the first block in Select mode. 

One thing I'm noticing is that focusing on resize handle and pressing left arrow once sometimes takes me into edit mode straightaway. I'm not sure if that's desirable, as currently the drag handle, when dragged with a mouse, allows us to expand the canvas until the sidebar is about 100px wide, and only if we try expanding further than that does it switch to edit mode. Dragging with a mouse is also behaving a bit weirdly in this branch:


https://github.com/WordPress/gutenberg/assets/8096000/3c6d2ca4-15d6-47b8-a427-ff221bfcf484

(I can't reproduce this 100% of the time, but it's happening consistently on Firefox and about half the time on Chrome)

**Review by tellthemachines — COMMENTED** [review]
Thanks for fixing this! It's generally working well, apart from what I mentioned below.

One thing (unrelated to this PR) I noticed when testing is that there's no way to skip directly into edit mode with the keyboard; we have to tab through the whole sidebar to get to the canvas frame. Would be nice to have a skip link or something at the start.

**Comment by andrewhayward** [comment]
Couple of thoughts...

## Edit mode

First, it's really jarring and unexpected to jump into edit mode by hitting left one too many times, and I'd fail it under [WCAG 3.2.2](https://www.w3.org/TR/WCAG21/#on-input):

> Changes in context can confuse users who do not easily perceive the change or are easily distracted by changes. Changes of context are appropriate only when it is clear that such a change will happen in response to the user's action.
> — [Understanding SC 3.2.2: On Input (Level A)](https://www.w3.org/WAI/WCAG21/Understanding/on-input.html)

I'd suggest instead that we clamp the resizing to the maximum and minimum sizes available, and leave it as a simple resize control.

<details>
  <summary>Diff</summary>

```diff
diff --git a/packages/edit-site/src/components/resizable-frame/index.js b/packages/edit-site/src/components/resizable-frame/index.js
index 22567dbe96..cb5f5120ee 100644
--- a/packages/edit-site/src/components/resizable-frame/index.js
+++ b/packages/edit-site/src/components/resizable-frame/index.js
@@ -176,9 +176,12 @@ function ResizableFrame( {
 
 		const step = 20 * ( event.shiftKey ? 5 : 1 );
 		const delta = step * ( event.key === 'ArrowLeft' ? 1 : -1 );
-		const newWidth = Math.max(
-			FRAME_MIN_WIDTH,
-			frameRef.current.resizable.offsetWidth + delta
+		const newWidth = Math.min(
+			Math.max(
+				FRAME_MIN_WIDTH,
+				frameRef.current.resizable.offsetWidth + delta
+			),
+			initialComputedWidthRef.current
 		);
 
 		setFrameSize( {
@@ -188,14 +191,6 @@ function ResizableFrame( {
 				initialAspectRatioRef.current
 			),
 		} );
-
-		// If oversized, immediately switch to edit mode
-		if ( newWidth > initialComputedWidthRef.current ) {
-			setFrameSize( INITIAL_FRAME_SIZE );
-			setCanvasMode( 'edit' );
-			// TODO: Confirm if this is the focus behavior we want
-			document.querySelector( 'iframe[name=editor-canvas]' ).focus();
-		}
 	};
 
 	const frameAnimationVariants = {
```
</details>

## Component role

Secondly, we should ideally mark the resize handle as a `separator`, and (because it's interactive) provide appropriate `aria-value*` attributes.

```html
<el role="separator" aria-valuemin="..." aria-valuemax="..." aria-valuenow="..." />
```

<details>
  <summary>Diff</summary>

```diff
diff --git a/packages/edit-site/src/components/resizable-frame/index.js b/packages/edit-site/src/components/resizable-frame/index.js
index 22567dbe96..e500723c31 100644
--- a/packages/edit-site/src/components/resizable-frame/index.js
+++ b/packages/edit-site/src/components/resizable-frame/index.js
@@ -231,6 +231,10 @@ function ResizableFrame( {
 		return shouldShowHandle ? 'visible' : 'hidden';
 	} )();
 
+	const maxSize = initialComputedWidthRef.current;
+	const minSize = FRAME_MIN_WIDTH;
+	const currentSize = frameRef?.current?.resizable?.offsetWidth || maxSize;
+
 	return (
 		<ResizableBox
 			as={ motion.div }
@@ -281,8 +285,13 @@ function ResizableFrame( {
 								) }
 								variants={ resizeHandleVariants }
 								animate={ currentResizeHandleVariant }
+								role="separator"
+								aria-orientation="vertical"
 								aria-label={ __( 'Drag to resize' ) }
 								aria-describedby={ resizableHandleHelpId }
+								aria-valuenow={ currentSize }
+								aria-valuemin={ minSize }
+								aria-valuemax={ maxSize }
 								onKeyDown={ handleResizableHandleKeyDown }
 								initial="hidden"
 								exit="hidden"
```
</details>

**Comment by tellthemachines** [comment]
We should aim to have this wrapped up before the end of the week so it can make 6.3 RC1. Do you think that's possible @mirka?

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (afercia) [reply packages/edit-site/src/components/resizable-frame/index.js]: Hello @mirka @tellthemachines I'd agree focus management should work in the same way it works when switching to the Edit view when pressing Enter on the iframe. However, I'm not sure it should land on the first block... I created https://github.com/WordPress/gutenberg/issues/51570 a few days ago to reconsider initial focus on the Site editor > Edit view.

**Comment by mirka** [comment]
> We should aim to have this wrapped up before the end of the week so it can make 6.3 RC1. Do you think that's possible @mirka?

Yes, I'll try to land this ASAP 🙏

**Comment by afercia** [comment]
Quickly tested and seems to me it works pretty well, thank you!

One thing I found a bit confusing is that the 'Drag to resize' handle is placed in the tab sequence _after_ the iframe. I would expect to find it between the navigation and the iframe so that it also matches the visual order.

A few more things:
- The handle button needs a `role="separator"`. This would also help interaction when using a screen reader, as left and right arrow keys may be reserved to screen reader usage especially with Windows screen readers. Not sure if that applies to button elements though. Cc @alexstine. See [the role=separator in the ARIA spec](https://www.w3.org/TR/wai-aria-1.2/#separator). 
- Worth reminding that in ARIA terms this is a 'window splitter' widget. It is described at https://www.w3.org/WAI/ARIA/apg/patterns/windowsplitter/
- Ideally, if there is a way to calculate the value of the resizing, we should use that value for an `aria-valuenow` attribute to be added to the handle button. See https://www.w3.org/TR/wai-aria-1.2/#aria-valuenow

**Comment by afercia** [comment]
I forgot the separator should also have an `aria-orientation="vertical"` attribute ❤️

**Comment by annezazu** [comment]
Flagging for @alexstine and @joedolson in case you all want to help review too <3

**Comment by joedolson** [comment]
I feel like @afercia's comments are pretty thorough. There's nothing there I don't agree with.

**Comment by annezazu** [comment]
Awesome 🥳 Just wanted to make sure these feedback chances are visible with the release winding up.

**Comment by mirka** [comment]
Thanks for the feedback, everyone! This is now ready for final review.

I've addressed the two main issues:

- Semantically make the resize handle a `separator`. Appropriate aria attributes have been added as well.
- Hitting the left arrow key on the separator should not make it go into Edit mode, as it is unexpected and jarring.


Tangential issues that have been deferred from this PR:

- The mouse dragging glitch that @tellthemachines reported. Unfortunately I could not reproduce in Firefox nor Chrome. Though if it is a bug I feel like it should be reproducible in trunk as well (i.e. not a regression) since the mouse dragging bits aren't really changed in this PR. We could look into it separately.
- The tab sequence order of the resize handle, as raised by @afercia. This is not a quick fix so we should address it separately. #52602
- #51570
- #52515
- #52516

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (tellthemachines) [inline packages/edit-site/src/components/resizable-frame/index.js:182]: Might be good to include something about using Shift to jump five steps in the message below?

**Review by tellthemachines — COMMENTED** [review]
Code looks good, just one minor note below!

One thing I notice on testing is that when focus lands on the resizer, screen readers announce the "you are currently in navigation mode..." spiel in addition to the "separator resizer use left and right arrow keys...". I tested with VO/Safari and NVDA/Firefox. With NVDA it's a bit worse because the navigation mode bit is the first thing you hear when landing on the button, whereas with VO it comes after the relevant instructions.

Another thing is that when tabbing through from the sidebar, there's an invisible tab stop between the "Saved" button and the canvas frame, though that's also on trunk so not specific to this PR. It's also a regression from 6.2 though.

**Comment by afercia** [comment]
@tellthemachines regarding the last two points you reported (good catch), I alraedy created two issues ~ one month ago. It would be great it they could be triaged.

> when focus lands on the resizer, screen readers announce the "you are currently in navigation mode..

See https://github.com/WordPress/gutenberg/issues/51578

> when tabbing through from the sidebar, there's an invisible tab stop between the "Saved" button and the canvas frame

See https://github.com/WordPress/gutenberg/issues/51540

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (afercia) [inline packages/edit-site/src/components/resizable-frame/index.js]: If it makes things easier, we can keep this as a button element. The ARIA role will override the button semantics anyways. Note the ARIA Authoring Practices example uses a div only for demoing purposes.

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (afercia) [inline packages/edit-site/src/components/resizable-frame/index.js]: Redundant if we keep this as a button.

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (afercia) [inline packages/edit-site/src/components/resizable-frame/index.js:295]: I'm thinking aria-valuemax should always be the main document offsetWidth. Otherwise this value will be different depending on the editor mode. For example: My screen is a 16-inch retina display. The default scaled resolution is 1792 x 1120 so the actual viewport width is 1792. WHen the editor is in vide mode, this will be aria-valuemax="1408". When the editor is in edit mode, this will be aria-valuemax="1792".
I think the value should always the actual maximum width the iframe can be set to, which is 1792.
Also, we should make sure this value updates when resizing the browser window (I thikn it does update when set to the main document offsetWidth).

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (afercia) [inline packages/edit-site/src/components/resizable-frame/index.js:172]: Please add `event.preventDefault();` here. This should help prevent screen readers from reading the button text description character by character when pressing the Right or Left arrow keys.

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (afercia) [inline packages/edit-site/src/components/resizable-frame/index.js]: I'd use a hidden div `<div hidden ...` instead of a `VisuallyHidden` component, to make sure the description text doesn't get read by screen readers when pressing the Right or Left arrow keys. The aria-describedby association works also when it targets a hidden element.

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (afercia) [inline packages/edit-site/src/components/resizable-frame/index.js:179]: Please add `event.preventDefault();` here. This should help prevent screen readers from reading the button text description character by character when pressing the Right or Left arrow keys.
Edit: make sure to place it _after_ the check for arrow keys because other keys e.g. Tab still needs to work as expected.

**Comment by afercia** [comment]
> The tab sequence order of the resize handle, as raised by @afercia. This is not a quick fix so we should address it separately.

Is there a follow-up issue for this already? I don't want to block this PR so yes it can be addressed separately. But I'd be a bit uncomfortable with releasing this feature in WP 6.3 with a confusing tab sequence.

**Comment by afercia** [comment]
On a side note:
Why the minimum width is 340? It appears to be a magic number. Is there any reason for this value? I'm not sure 340 pixels is a common device viewport width. If the goal is to allow users to emulate the most common mobile viewport widths, then maybe a more useful value would be 320?

Edit:
Worth also noting that when clicking Styles > Styles Book, the editor switches in 'focus mode' and it'r resizable. In this mode, [the minimum width is 300](https://github.com/WordPress/gutenberg/blob/60e3299ab33f024b0d1d36e4c097d03dabf58c75/packages/edit-site/src/components/block-editor/resizable-editor.js#L45). Not sure why it must be 300 in focus mode and 340 for the iframe. I'd consider to standardize to the same minimum width.

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (mirka) [reply packages/edit-site/src/components/resizable-frame/index.js]: Good to know, thanks 👍 5ccb4b9

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (mirka) [reply packages/edit-site/src/components/resizable-frame/index.js]: Good point, that is an important distinction. 96f95c6

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (mirka) [reply packages/edit-site/src/components/resizable-frame/index.js:179]: Done 👍 f5b7a8f

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (mirka) [reply packages/edit-site/src/components/resizable-frame/index.js:182]: Good idea! f5bbecd

File: packages/edit-site/src/components/resizable-frame/index.js
Comment (mirka) [reply packages/edit-site/src/components/resizable-frame/index.js:295]: This is a difficult one actually. The values are associated with the separator, and the separator is only functional in View mode. Especially in a keyboard operation context, where I believe the semantics are most relevant, the max value is in fact the initial computed width. I think it would be confusing if you couldn't keyboard resize the frame up to the purported max value.

(In testing this, I noticed that the resize handle is still in the tab sequence when in Edit mode, so I fixed that. 5acbbfd)

**Comment by mirka** [comment]
@afercia 

> Is there a follow-up issue for this already? I don't want to block this PR so yes it can be addressed separately. But I'd be a bit uncomfortable with releasing this feature in WP 6.3 with a confusing tab sequence.

I will post an issue after this PR is merged. However, this is a tricky one, because in the general case, resize handles can be on any edge or corner. It's only in this specific case where the handle is only on the left edge. So I do see an argument to optimize for consistency in the general case (for example the one in Style Book with two handles), and have all handles come after the frame.

> If the goal is to allow users to emulate the most common mobile viewport widths, then maybe a more useful value would be 320?

That makes sense. I changed it to 320px ([5fa0e8d](https://github.com/WordPress/gutenberg/pull/52443/commits/5fa0e8d3f442be951ac1c7ceb1b90c708ae747f7)), and I think we can change the Style Book one as well.

**Comment by mirka** [comment]
@tellthemachines I just realized that the dragging glitch you were seeing is likely #52582. I'll try to get that patched ASAP as well.

**Review by tellthemachines — APPROVED** [review]
Thanks for the follow-ups @mirka ! I think this is in a good place to be merged now, to fix the major regression from 6.2. Then we can address the rest of the issues.

One small thing that could be improved: screen readers announce the resizing in percentages but I'm not sure that makes much sense for the purpose. Wouldn't it be more useful to know the current canvas size in px? In any case, given the 6.2 resize handle doesn't announce anything, that's fine to iterate on for 6.4.

**Comment by mirka** [comment]
> Wouldn't it be more useful to know the current canvas size in px?

Yes, we can do that 👍

**Comment by tellthemachines** [comment]
I just cherry-picked this PR to the update/first-batch-pre-RC1 branch to get it included in the next release: 4df1030888

**Comment by afercia** [comment]
Thanks for working on this ❤️ 

> Wouldn't it be more useful to know the current canvas size in px?

I'm not sure that would be what screen readers expect. The value of [aria-valuenow is expected to be a decimal number](https://www.w3.org/TR/wai-aria-1.2/#aria-valuenow). No pixels please. Once screen readers are provided with correct values for aria-valuenow, aria-valuemin, and aria-valuemax, then they are able to do their own calculations and provide users with some meaningful announcement. I don't think the ARIA specification tells anything about the 'format' of the announcement, this is left to each vendo implementation. We should just provide the values and let screen readers do their job.

Theoretically, [aria-valuetext](https://www.w3.org/TR/wai-aria-1.2/#aria-valuetext) could be used to override the screen reader behavior and provide a human readable text alternative of [aria-valuenow](https://www.w3.org/TR/wai-aria-1.2/#aria-valuenow), Seems a bit overkill to me though.

**Pull request merged**
